### PR TITLE
match-mode

### PR DESCRIPTION
**Pull request by kurtbrose**

### CODE REVIEW

File: glom/matching.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+"""
+from boltons.typeutils import make_sentinel
+
+import weakref
+
+from .core import GlomError, glom, T, Spec
+
+
+class GlomMatchError(GlomError): pass
```
Comment (mahmoud) [inline glom/matching.py]: relatively big todo here: we need the path and so forth to craft actionable error messages.

File: glom/test/test_match.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    glom(1.0, (M > 0) & float)
+    glom(1.0, (M > 100) | float)
+    glom(1.0, M & (M > 0) & float)
+    glom(1.0, M | (M > 100) | float)
+    # test idiom for enum
+    with pytest.raises(GlomMatchError):
+        glom("c", Match("a"))
+    with pytest.raises(GlomMatchError):
+        glom("c", Match(Or("a", "b")))
+    _chk(Match(M | "a" | "b"), "a", "c")
```
[the remark this answers is not in the document]
Comment (mahmoud) [reply glom/test/test_match.py]: Another thing that popped into mind, `M` probably doesn't have any direct comparison usage for callability, so we could make that a shortcut for `Match` construction, maybe?

File: glom/core.py [hunk 3]
```diff
@@ -1133,6 +1133,57 @@ def glomit(self, target, scope):
         return ret
 
 
+class Capture(object):
```
Comment (mahmoud) [inline glom/core.py]: capture is really its own thing right? doesn't seem to directly influence match?

File: glom/matching.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+    # TODO: straightforward to extend this to all comparisons
+
+    def glomit(self, target, scope):
+        lhs, op, rhs = self.lhs, self.op, self.rhs
+        if lhs is M:
+            lhs = target
+        if rhs is M:
+            rhs = target
+        if op == '=':
```
[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/matching.py]: maybe silly, but I have vague ideas that there is an optimization cliff around 1 character strings

[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/matching.py]: of course `>=` etc will be SUPER awkward to jam into a single character so maybe I should abandon it

File: glom/matching.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+                raise GlomMatchError("{!r} > {!r}".format(lhs, rhs))
+        return target
+
+    def __repr__(self):
+        if self is M:
+            return "M"
+        op = self.op
+        if op == '=':
+            op = '=='
+        return "{} {} {}".format(repr(self.lhs), op, repr(self.rhs))
```
[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/matching.py]: 👍 will do

File: glom/matching.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+ensure that dictionary values are all strings
+Match({DEFAULT: str})
+
+2 syntaxes for combining expressions:
+
+And(int, M > 0)
+M & int & (M > 0)
+
+"""
+from boltons.typeutils import make_sentinel
```
[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/matching.py]: 👍 will do

File: glom/test/test_match.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/test/test_match.py]: I was thinking maybe the name of `M` should actually change to `C` for "chainable comparator" or something

Basically, `M` and `Match` aren't necessarily connected in usage or implementation even though they arose close to each other

I think keeping `M` and `Match` distinct is really good because `M` is useful inside other schemes/dialects, and also it eliminates a `Literal` vs `Spec` problem

[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/test/test_match.py]: but, I definitely had the exact same thoughts around -- should the public constructor for `M` be an alias for `Match()`?

[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/test/test_match.py]: (and have not at all rejected that idea)

File: glom/matching.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+M = MType(None, None, None)
+
+
+def _precedence(match):
+    """
+    in a dict spec, target-keys may match many
+    spec-keys (e.g. 1 will match int, M > 0, and 1);
+    therefore we need a precedence for which order to try
+    keys in; higher = later
+    """
```
[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/matching.py]: ooooooh!!!!! good idea!

File: glom/matching.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+                try:
+                    _glom_match(item, child, scope)
+                    break
+                except GlomMatchError as e:
+                    last_error = e
+            else:  # did not break, something went wrong
+                if target and not spec:
+                    raise GlomMatchError("{!r} does not match empty {}".format(
+                        target, type(spec).__name__))
+                raise last_error
```
[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/matching.py]: ooooooooppppss

File: glom/matching.py (hunk 1 above)
Comment (kurtbrose) [reply glom/matching.py]: agree!

File: glom/core.py (hunk 3 above)
Comment (kurtbrose) [reply glom/core.py]: yes this could (should?) be broken into it's own thing

Comment (kurtbrose) [reply glom/core.py]: capture is in the area of pattern matching, basically it is adjacent to the match concept so I was kind of prototyping it in the same place

**Comment by kurtbrose** [comment]
thanks for the comments, I've got some stuff to address;

it's at the point now where if the shape of the API seems correct I can go through and tighten things down

File: glom/matching.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mahmoud) [reply glom/matching.py]: haha that hadn't occurred to me, and personally I'd probably abandon it yeah.

File: glom/matching.py [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+                # so last_error will have been assigned
+                raise last_error
+    elif isinstance(spec, tuple):
+        if not isinstance(target, tuple):
+            raise GlomTypeMatchError(type(target), tuple)
+        if len(target) != len(spec):
+            raise GlomMatchError("{!r} does not match {!r}", target, spec)
+        for sub_target, sub_spec in zip(target, spec):
+            scope[glom](sub_target, sub_spec, scope)
+    elif isinstance(spec, type):
```
[the remark this answers is not in the document]
Comment (mahmoud) [reply glom/matching.py]: Ah right, the whole should it act as a converter or a validator debate.

File: glom/matching.py [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+    if callable(match):
+        return 1
+    return 0  # == match
+
+
+def _handle_dict(target, spec, scope):
+    if not isinstance(target, dict):
+        raise GlomTypeMatchError(type(target), dict)
+    spec_keys = spec  # cheating a little bit here, list-vs-dict, but saves an object copy sometimes
+    if sys.version_info < (3, 6):
```
[the remark this answers is not in the document]
Comment (mahmoud) [reply glom/matching.py]: (We could also run our own test, throw a few dozen items into a dict at import time, see if they come out the same order?)

File: glom/matching.py (hunk 1 above)
Comment (kurtbrose) [reply glom/matching.py]: short term: is there something to do to make `__repr__` kind of work as intended?  (without doing more upfront work than constructing a tuple)

File: glom/matching.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/matching.py]: I'm a little nervous about testing ordering accidentally working on certain platforms, but 100% agree we should use a good "is_sorted_dict"

hey!  does that sound like a bolton to you?  `dict_utils.DICT_IS_ORDERED`?

File: glom/matching.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/matching.py]: yea, I think we need to re-learn the idioms for using this -- because `Auto` mode is right next door, it doesn't have to take on as much itself

[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/matching.py]: check out `test_matching.test_nested_struct()`

[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/matching.py]: check out `test_matching.test_nested_struct()`

[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/matching.py]:
```python
    _json = lambda spec: Auto((json.loads, Match(spec)))

    _defaults = lambda **kw: Merge(Fill([T]), init=lambda: dict(kw))

    rule_spec = (
        _defaults(save_as_new=False),
        Match({
            'rule_id': Or('', Regex(r'\d+')),
            'rule_name': basestring,
            'effect': Or('approve', 'custom_approvers'),
            'rule_approvers': _json([{'pk': int, 'level': int}]),
            'rule_data': _json([  # list of condition-objects
                Auto((
                    _defaults(value='null'),
                    Match({  
                        'value': _json(
                            Or(None, int, float, basestring, [int, float, basestring])),
                        'field': Auto(int),  # id of row from FilterField
                        'operator': basestring,  # corresponds to FilterOperator.display_name
                    })))]),
            'save_as_new': Or(basestring, bool),
        }))
```

File: glom/matching.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mahmoud) [reply glom/matching.py]: Sure, it can go into boltons. Also, a hundred random items going into a dict and coming out the same exact order, that'd be mighty strange yeah?

**Comment by kurtbrose** [comment]
got another use case -- what if you want to compare sub-trees?

context, you want to verify a bunch of leaf sums add up to the root sum

```python
(Let(leaf_sum=Sum( ('leaves', ['val'])),
'total',
M == S['leaf_sum']
)
```

this is awkward and doesn't technically work since `M ==` will always do a literal comparison

also, it leaves the target cursor pointed at 'total', whereas it is probably nicer to leave the target cursor in place

a nicer way to write this out would be

```python
M('total') == M(Sum( ('leaves', ['val']) ))
```
or similar

**Comment by kurtbrose** [comment]
implementation of above would look something like

```
class MType:
    # ...
    def __call__(self, spec):
        return _SpecChild(spec)

    def glomit(self, target, scope):
        # ...
        if type(lhs) is _SpecChild:
           lhs = scope[glom](target, lhs.spec, scope)
        if type(rhs) is _SpecChild:
           rhs = scope[glom](target, rhs.spec, scope)
        # ...
```

**Comment by kurtbrose** [comment]
the `M(subspec)` above use case looks like another case of the Literal vs Spec corner case

**Comment by kurtbrose** [comment]
```python
def test_spec_match():
    """test that M __call__ can be used to wrap a subspec for comparison"""
    target = {}
    target['a'] = target
    assert glom(target, M == M('a'))
```

**Comment by kurtbrose** [comment]
a major milestone has been reached -- finally able to express tree transformations:

```python
def test_json_ref():
    assert glom(
        {'a': {'b': [0, 1]}},
        Ref('json',
            Match(Or(
                And(dict, {Ref('json'): Ref('json')}),
                And(list, [Ref('json')]),
                And(0, Auto(lambda t: None)),
                object)))) == {'a': {'b': [None, 1]}}
```

I guess in fairness to compare it to equivalent python:

```python
def _0toNone(val):
    if type(val) is dict:
        return {_0toNone(key): _0toNone(val) for key, val in val.items()}
    if type(val) is list:
        return [_0toNone(item) for item in val]
    if val == 0:
        return None
    return val
```

File: glom/matching.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/matching.py:594]: I could imagine some implementation deciding that linear search is faster than a real table maybe?  (since it is mostly going to be doing integer comparisons on a hash)

I guess it will become less and less relevant what mechanism we use over time

[the remark this answers is not in the document]
Comment (kurtbrose) [reply glom/matching.py:594]: (since I expect sorted to become more and more the default)

**Comment by mahmoud** [comment]
Next steps:

- [ ] * Add some "Note" asides to the docs where necessary
- [ ] * Add better exception docstrings
- [ ] * Maybe split up various MatchError types (though maybe not, because we may be revisiting this for multi error)
- [ ] * Maybe move Check into matching module
- [ ] * Change CheckError aside to a recommendation to use Match
- [ ] * Add a default to Match() (like Check())
- [ ] * Make sure Match() covers all Check() cases (at least in the tests)
- [ ] * Add a `default` argument to Match (like Check)
- [ ] * Add a `glom.match()` top-level function.

**Comment by kurtbrose** [comment]
just remembered: waaaaaay back I got a request for support for typing constants --

for whatever reason some people find them more readable in a static context, I see no reason why not to support them

```python
typing.Dict[str, int]  # Match() treats the same as {str: int}
typing.List[float]  # Match() treats the same as [float]
typing.Tuple(str, int)  # Match() treats the same as (str, int)
```

its just a little bit of extra dispatch logic to support those

there might be something interesting around e.g. `Coroutine`, `Callable`, etc in typing

don't want to hold up the merge -- maybe this will lift into a follow on issue

**Comment by kurtbrose** [comment]
typing `Callable` might be an interesting way to match on function arg signature..... although it doesn't reflect arg names

https://docs.python.org/3/library/typing.html#typing.Callable

File: docs/snippets.rst [hunk 11]
```diff
@@ -233,6 +233,29 @@ with a dict:
...
+
+    glom(range(10), [(M < 7) | Literal(7)])
+    # [0, 1, 2, 3, 4, 5, 6, 7, 7, 7]
+
+
+What if you want to drop rather than clamp out-of-range values?
+
+.. code-block:: python
+
+    glom(range(10), [(M < 7) | Literal(SKIP)])
```
Comment (mahmoud) [inline docs/snippets.rst:255]: we should make it such that it's not necessary to add Literal() around SKIP. Either by making it an instance of a type with a glomit or having M evaluation build special handling for it.

Comment (mahmoud) [reply docs/snippets.rst:255]: As in, the end-user of glom should never have SKIP or STOP in their datastructures, and they should never end up with it being in the result.

**Comment by mahmoud** [comment]
idea i had (probably for a follow-on): `M` (and related types) should probably implement `__bool__`. It's basically doing a truthy check on its own, so `bool(M)` would be a noop more or less. And then for cases like `bool(And(M ==1, M < 2))` it'd pull it back to True/False.

File: docs/snippets.rst (hunk 11 above)
Comment (kurtbrose) [reply docs/snippets.rst:255]: sounds good to me -- to be precise the problem here isn't SKIP being emitted in the output, it is `SKIP` not being valid as a glom

we could fix this by having

```python
def glomit(self, target, scope):
   return self
```

attached to `SKIP` somehow

**Comment by kurtbrose** [comment]
should dicts and tuples in match mode respect SKIP and STOP?

**Comment by kurtbrose** [comment]
I think SKIP and STOP being implemented via exception will make this come more naturally

**Comment by kurtbrose** [comment]
things that occur reading the new docs  https://glom.readthedocs.io/en/match-mode/

in no particular order

- [ ] Iter() reference Auto(list) as another option
- [ ] Exceptions also explain stack-trace-formatting
- [ ] public docs for Fill
- [ ] public docs explaining a "Mode" (not just for extending glom, but also for users)
- [ ] mark glom.Glommer() as deprecated
- [ ] public docs for Group (might be helpful to explain as a combination of Iter and Flatten)
- [ ] rewrite JSON tree walk snippet to use Switch once merged

**Comment by kurtbrose** [comment]
* [ ] kurt to do bb-repr in match error formats
